### ISSUE TITLE
nvim__redraw causes some visible screen tearing

### ISSUE THREAD

**Issue by gpanders** [neovim/neovim#28668]
### Problem

Since https://github.com/neovim/neovim/commit/037ea6e786b5d05f4a8965e4c2ba6aa60ec7c01a, Nvim occasionally exhibits some black "tearing"/flickering on startup.

It is difficult to capture so I recorded a slow motion video on my phone:

**Before (no black tearing)**:
 
https://github.com/neovim/neovim/assets/8965202/ffd619d5-7714-43be-8387-409b80962ff7

**After (with black tearing)**:

https://github.com/neovim/neovim/assets/8965202/468c8800-2d34-4003-b7bb-975ddb288905

Note that in both cases we can see the screen state being drawn incrementally, but in the latter video the background of the top half of the screen becomes black first. In real time this has a noticeable "flicker" effect.

### Steps to reproduce

In my case this seems to be caused by the `nvim__redraw` call in the Tree-sitter highlighter, so the following is sufficient to reproduce the issue:

```
cat >test.vim <<EOF
colorscheme blue
au FileType c lua vim.treesitter.start()
EOF
nvim --clean -u test.vim src/nvim/main.c
```

### Expected behavior

The background should not become black when redrawing.

### Neovim version (nvim -v)

064f3e42e8d33b0a9e560dfb7c9a42b2fc1ed868

### Vim (not Nvim) behaves the same?

No

### Operating system/version

macOS

### Terminal name/version

Ghostty

### $TERM environment variable

xterm-ghostty

### Installation

Source

**Comment by gpanders** [comment]
cc @luukvbaal

**Comment by gpanders** [comment]
The specific cause seems to be the `ui_flush()` call at the end of `nvim__redraw()`, which was not present in `nvim__buf_redraw_range()`.

**Comment by zeertzjq** [comment]
Hmm, it doesn't make sense to call `ui_flush()` if `update_screen()` isn't called.

**Comment by gpanders** [comment]
Fixed by #28676.
